### ISSUE TITLE
Bundle dependencies not really complete

### ISSUE THREAD

**Issue by jeremyrickard** [deislabs/porter#313]
When we resolve dependencies, we're reading the manifest (porter.yaml) from the dependent bundle and that is sufficient for cases where the mixins within the bundle are not referencing anything that needs to be within the invocation image. For example, the helm mixin in a bundle dependency could reference a chart in a chart registry or the Azure mixin could use one of the built in types. 

This becomes insufficient when the mixin references something that needs to be in the invocation image. For example, the case when someone includes a chart that is not in a registry that can be fetched at runtime or if someone provides ARM templates to the Azure mixin or Terraform manifests to the Terraform mixin. In these cases, the current dependency capabilities would not include those and the dependency wouldn't achieve the desired capability. We do copy the bundle directory, but it's not really clear how we should distribute those bits as part of the bundle and handle the pathing. The manifests might be referenced in the porter.manifest with a fully qualified path (i.e. /cnab/app/manifests/terraform) but we are just copying the bundle directory into the bundle deps. 

I think we should revisit how dependencies need to be handled within Porter, or we need to write really explicit guidance or instructions about how they _should_ work. Having it highly constrained makes me worry about the experience.

**Comment by carolynvs** [comment]
The current state wasn't designed to be "highly constrained", it's just that way because we implemented it in the 3 days before Kubecon. 😀  Gosh it's all super fuzzy now but dependencies was just put together enough to show what we wanted to do at high level and [not finished](https://github.com/deislabs/porter/issues/109) so please don't assume intent behind how it works today. Everything is changable and as always a great developer experience is the goal. So thank you for bringing up that it sucks right now! This is our top priority on the [roadmap](https://github.com/deislabs/porter#roadmap) to showing the vision of Porter, so we want to get it right.

> This becomes insufficient when the mixin references something that needs to be in the invocation image.

I wanted to clarify the steps of what porter does when we depend on another bundle, to help us anchor the conversation in case anyone else is playing along at home 😀 

1. Load the root bundle's manifest, and discover a dependency.
2. Load the dependency's manifest and merge it into the root manifest. 
3. Start `porter build`, using the template Dockerfile if one was provided. FYI, we have a gap as of yesterday, now that a dependency can specify a template that we need to merge. But that isn't what you were referencing in this issue as the immediate concern.
4. All of the mixins are executed (`mixin build`) to declare what should be included in the the Dockerfile, **completely unaware of if they being run as a part of a root bundle or a dependency.** They are treated the same.
5. Porter copies the root bundle's porter.yaml into the /cnab/app/ directory. It also copies the dependencies into the /cnab/app/bundles, bringing any files that they distribute along.

At runtime, porter does the same set of steps to merge the manifests, and execute the steps defined in the manifest using the mixins.

The problem that quote is referencing (Jeremy correct me if I am wrong or missed other problems!), is **step 4 at runtime**. Porter took commands out of a dependent bundle that could have referenced files in that bundle, and now when the command executes the files aren't where the command expects.

## Example

Bundle A has a porter.yaml _and_ a script that it calls during install.

```
mixins:
- exec

install:
  - exec:
      description: "Install Hello World"
      command: bash
      arguments:
        - ./install-stuff.sh
```

Bundle B depends on A.

```
dependencies:
- name: A
```

When porter builds this into an invocation image this is file system:

```
cnab/
└── app
    ├── bundles
    │   └── A
    │       ├── install.sh
    │       └── porter.yaml
    ├── mixins
    │   └── exec
    │       ├── exec
    │       └── exec-runtime
    ├── porter-runtime
    ├── porter.yaml
    └── run
```

When the install step runs for bundle A it is going to run in /cnab/app and when it tries to call `./install.sh`, the call will fail because the file is located at `/cnab/app/bundles/A/install.sh`.

## Proposed Solution
Use the existing context that we keep around on the step when we execute it. Porter does know the dependency that owns the step its about to run, `Step.dep`. Instead of running the step in `/cnab/app`, what if we ran it in `/cnab/app/bundles/DEP`?

We would need to document that you cannot assume that the working directory of your mixin will always be the same, but it will always be in the directory of your bundle's manifest, next to your files.

What do you think? Would that be sufficient or do you see more gaps that we'd need to address around the problems created by moving commands out of bundle and into another?

**Comment by jeremyrickard** [comment]
In the example you gave, you're referencing something that's parallel to the manifest, is that something that would work as is? Prior to the the addition of the dockerfile template,  `porter build` would only add the `cnab` directory into image, along with the porter.yaml. To reference it as ./install.sh you'd need to know that when things are running you need to drop those files into /cnab/app so you can reference them relatively or you end up using the fully qualified path.  When ever we fix the merge of the dockerfile templates, I guess it would be possible to write into the dockerfile template to add individual files and assume they're going to be local, although that seems like it could get rather complex pretty quickly. 

Maybe it makes sense for us to have "known" places to copy stuff from? Maybe `porter create` also creates a directory called "stuff" and `porter build` will copy `stuff` into /cnab/app/stuff in the normal case, or /cnab/bundle/DEP/stuff in the dep case and you can assume ./stuff/file when you run things? Or is it fine to assume that people will just know they should put htings in /cnab/app and we'll copy everything from the dependency (except the run script?)

**Comment by carolynvs** [comment]
After much thinking and 🥞 I realized that our current path is unworkable with dependencies once we added in custom FROM (#311) yesterday. I sealed our fate. ☠️ 

Then I had an epiphany. 💡 We only went down this path of merging manifests anyway because the CNAB spec originally didn't have any way for us to execute a dependency and get anything out of it. Now that your outputs proposal is about to be merged, we can go back to what we originally wanted to do all along: execute the dependent bundle, retrieve the outputs from it, and smoosh them into `bundle.dependencies.mysql.outputs.host`.

I'm going to open up a new epic issue to consolidate the new direction for dependencies. I think this could be closed now but up to you if you'd prefer to keep this open until that work is complete.
